File: packages_py/app_yaml_config/src/app_yaml_config/resolve_auth.py

from dataclasses import dataclass
from typing import Dict, Any, Optional, Literal
from fetch_auth_config import fetch_auth_config, AuthConfig
from fetch_auth_encoding import encode_auth


@dataclass
class AuthResolutionResult:
    """Result of resolving authentication for a provider."""
    auth_config: AuthConfig
    headers: Dict[str, str]
    provider_name: str
    source: Literal['provider_config', 'env_overwrite', 'default']


@dataclass
class ResolveAuthOptions:
    """Options for auth resolution."""
    include_headers: bool = True


def resolve_provider_auth(
    provider_name: str,
    provider_config: Dict[str, Any],
    options: Optional[ResolveAuthOptions] = None
) -> AuthResolutionResult:
    """
    Resolve authentication configuration for a provider.
    Thin wrapper around fetch_auth_config that mirrors resolve_provider_proxy pattern.

    Args:
        provider_name: Name of the provider
        provider_config: Provider configuration dict
        options: Resolution options

    Returns:
        AuthResolutionResult with auth config, headers, and source info
    """
    opts = options or ResolveAuthOptions()

    # Use fetch_auth_config for credential resolution
    auth_config = fetch_auth_config(provider_name, provider_config)

    # Build headers if requested
    headers: Dict[str, str] = {}
    if opts.include_headers:
        creds = _build_credentials(auth_config)
        headers = encode_auth(auth_config.type.value if hasattr(auth_config.type, 'value') else str(auth_config.type), **creds)

    # Determine source based on resolution metadata
    source: Literal['provider_config', 'env_overwrite', 'default'] = 'provider_config'
    if auth_config.resolution and auth_config.resolution.resolved_from:
        if len(auth_config.resolution.resolved_from) > 0:
            source = 'env_overwrite'

    return AuthResolutionResult(
        auth_config=auth_config,
        headers=headers,
        provider_name=provider_name,
        source=source,
    )


def _build_credentials(auth_config: AuthConfig) -> Dict[str, Any]:
    """Build credentials dict from auth config."""
    creds: Dict[str, Any] = {}
    if auth_config.token:
        creds['token'] = auth_config.token
    if auth_config.username:
        creds['username'] = auth_config.username
    if auth_config.password:
        creds['password'] = auth_config.password
    if auth_config.email:
        creds['email'] = auth_config.email
    if auth_config.header_name:
        creds['header_key'] = auth_config.header_name
    if auth_config.header_value:
        creds['header_value'] = auth_config.header_value
    return creds

File: packages_mjs/app-yaml-config/src/resolve-auth.ts

import { fetchAuthConfig, AuthConfig, AuthConfigError } from '@internal/fetch-auth-config';
import { encodeAuth } from '@internal/fetch-auth-encoding';

export interface AuthResolutionResult {
    authConfig: AuthConfig;
    headers: Record<string, string>;
    providerName: string;
    source: 'provider_config' | 'env_overwrite' | 'default';
}

export interface ResolveAuthOptions {
    includeHeaders?: boolean;
}

/**
 * Resolve authentication configuration for a provider.
 * Thin wrapper around fetch-auth-config that mirrors resolve-proxy pattern.
 */
export async function resolveProviderAuth(
    providerName: string,
    providerConfig: Record<string, any>,
    options: ResolveAuthOptions = {}
): Promise<AuthResolutionResult> {
    const { includeHeaders = true } = options;

    // Use fetch-auth-config for credential resolution
    const authConfig = await fetchAuthConfig({
        providerName,
        providerConfig,
    });

    // Build headers if requested
    let headers: Record<string, string> = {};
    if (includeHeaders) {
        const creds = buildCredentials(authConfig);
        headers = encodeAuth(authConfig.type, creds);
    }

    // Determine source based on resolution metadata
    let source: AuthResolutionResult['source'] = 'provider_config';
    // Use optional chaining safely
    if (authConfig.resolution?.resolvedFrom) {
        const resolvedKeys = Object.keys(authConfig.resolution.resolvedFrom);
        if (resolvedKeys.length > 0) {
            source = 'env_overwrite';
        }
    }

    return {
        authConfig,
        headers,
        providerName,
        source,
    };
}

function buildCredentials(authConfig: AuthConfig): Record<string, any> {
    const creds: Record<string, any> = {};
    if (authConfig.token) creds.token = authConfig.token;
    if (authConfig.username) creds.username = authConfig.username;
    if (authConfig.password) creds.password = authConfig.password;
    if (authConfig.email) creds.email = authConfig.email;
    if (authConfig.headerName) creds.headerKey = authConfig.headerName;
    if (authConfig.headerValue) creds.headerValue = authConfig.headerValue;
    return creds;
}

// Re-export types from fetch-auth-config for convenience
export { AuthConfig, AuthType, AuthConfigError } from '@internal/fetch-auth-config';
